src/core/thread/network_diagnostic.cpp

```diff
@@ -707,6 +707,290 @@ void NetworkDiagnostic::HandleDiagnosticReset(Coap::Message &aMessage, const Ip6
+        OT_ASSERT(sizeof(aNetworkDiagTlv.mNetworkData) >= networkData.GetLength());
+        memcpy(aNetworkDiagTlv.mNetworkData, networkData.GetNetworkData(), networkData.GetLength());
+        aNetworkDiagTlv.mNetworkDataCount = networkData.GetLength();
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kIp6AddressList:
+    {
+        Ip6AddressListTlv &ip6AddrList = static_cast<Ip6AddressListTlv &>(tlv);
+
+        tlvTotalLength = sizeof(tlv) + tlv.GetLength();
+        VerifyOrExit(ip6AddrList.IsValid());
+        OT_ASSERT(sizeof(aNetworkDiagTlv.mIp6AddrList) >= ip6AddrList.GetLength());
+        VerifyOrExit(message.Read(offset + sizeof(ip6AddrList), ip6AddrList.GetLength(),
+                                  aNetworkDiagTlv.mIp6AddrList) == ip6AddrList.GetLength());
+        aNetworkDiagTlv.mIp6AddrCount = ip6AddrList.GetLength() / OT_IP6_ADDRESS_SIZE;
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kMacCounters:
+    {
+        MacCountersTlv macCounters;
+
+        tlvTotalLength = sizeof(MacCountersTlv);
+        VerifyOrExit(message.Read(offset, tlvTotalLength, &macCounters) == tlvTotalLength);
+        VerifyOrExit(macCounters.IsValid());
+        ParseMacCounters(macCounters, aNetworkDiagTlv.mMacCounters);
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kBatteryLevel:
+    {
+        BatteryLevelTlv batteryLevel;
+
+        tlvTotalLength = sizeof(BatteryLevelTlv);
+        VerifyOrExit(message.Read(offset, tlvTotalLength, &batteryLevel) == tlvTotalLength);
+        VerifyOrExit(batteryLevel.IsValid());
+        aNetworkDiagTlv.mBatteryLevel = batteryLevel.GetBatteryLevel();
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kSupplyVoltage:
+    {
+        SupplyVoltageTlv supplyVoltage;
+
+        tlvTotalLength = sizeof(SupplyVoltageTlv);
+        VerifyOrExit(message.Read(offset, tlvTotalLength, &supplyVoltage) == tlvTotalLength);
+        VerifyOrExit(supplyVoltage.IsValid());
+        aNetworkDiagTlv.mSupplyVoltage = supplyVoltage.GetSupplyVoltage();
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kChildTable:
+    {
+        ChildTableTlv &childTable = static_cast<ChildTableTlv &>(tlv);
+
+        tlvTotalLength = sizeof(childTable) + tlv.GetLength();
+        VerifyOrExit(childTable.IsValid());
+        VerifyOrExit(childTable.GetNumEntries() <= OT_ARRAY_LENGTH(aNetworkDiagTlv.mChildTable));
+        for (uint8_t i = 0; i < childTable.GetNumEntries(); ++i)
+        {
+            ChildTableEntry childEntry;
+            VerifyOrExit(childTable.ReadEntry(childEntry, message, offset, i) == OT_ERROR_NONE, error = OT_ERROR_PARSE);
+            ParseChildEntry(childEntry, aNetworkDiagTlv.mChildTable[i]);
+        }
+        aNetworkDiagTlv.mChildCount = childTable.GetNumEntries();
+        break;
+    }
+
+    case NetworkDiagnosticTlv::kChannelPages:
+    {
+        tlvTotalLength = sizeof(tlv) + tlv.GetLength();
+        OT_ASSERT(sizeof(aNetworkDiagTlv.mChannelPages) >= tlv.GetLength());
```

Comment:
Can we change the `OT_ASSERT()` into `VerifyOrExit()`?
We generally don't want to assert when parsing a received message (e.g.,  if the TLV from a received message is not well-formed and has bad/long length).

Reply:
I agree. This should be replaced with `VerifyOrExit` since we now allows extended TLV.